Yubo test case gen

yubo56 opened the pull request:
Wrote up a test_case_gen.py script for test cases 0, 1 using three flows Dummy, Reno, Fast

yubo56 committed:
test

yubo56 committed:
Added test case 0, 1 for flows Dummy, Reno, Fast

Added test case generation for all three using NetworkTopology
Also included constants in common.py for my use.

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+if __name__ == "__main__":
+    # test case 0
+    links = list()
+    l1 = Link("H1", "H2", h1, h2, LinkBuffer(64 * KILOBYTE), 0.01,
```
There are a lot of params in these constructors. You should use named variables to refer to each parameter. For example:

```
l1 = Link(end_1_addr="H1", end_2_addr="H2", ...)
```

Also, Links cannot have their end_1_device or end_2_device set. Please read the NetworkTopology preconditions. The JSON files generated by this script should be possible to feed into initializer and work successfully, which is not the case now. Please run this script and check that it works before submitting a new version.

yubo56 replied:
Sorry, didn't know about named variables :( Now passes initializer!

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    l1 = Link("H1", "H2", h1, h2, LinkBuffer(64 * KILOBYTE), 0.01,
+              10 * KILOBYTE, False)
+    links.append(l1)
+    h1.link = l1
```
h1 and h2 are undefined here, and should not have their links set anyways (see NetworkTopology preconditions).

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    flows.append(f1)
+    h1.flows["F1"] = f1
+
+    # (links, flows, hosts, routers) but we have no routers
```
This comment can be eliminated by using named parameters in function calls. See above.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+
+    # generate for FlowDummy
+    flows = list()
+    f1 = FlowDummy("F1", "H1", "H2", h1, h2, INITIAL_WINDOW_SIZE_PACKETS,
```
Use named parameters. Also, h1 and h2 should not be set here. From NetworkTopology's preconditions: "But no actual Host references should be connected to the Flows yet." You cannot store references in JSON so this will not work.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+"""
+Generates Test Cases 0, 1 and saves them to json
```
You should mention that these test cases are generated for each kind of Flow (FlowDummy, FlowReno, FlowFast).

You should also put all of the files in a data/ folder (where the "data" folder is in the same directory as this script). This should be mentioned here. Also, the format of the file names should be mentioned, i.e. "test_case_<num>_<flow_type>.json where num is ... and flow_types is ..."

lakshbhasin replied:
Also please add code for test case 2.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    f1 = FlowDummy("F1", "H1", "H2", h1, h2, INITIAL_WINDOW_SIZE_PACKETS,
+                                  set(), list(), 20 * MEGABYTE, 1)
+    flows.append(f1)
+    h1.flows["F1"] = f1
```
Do not set references at all in this script. This is handled in NetworkTopology's complete_initialization() function.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+
+    # (links, flows, hosts, routers) but we have no routers
+    network = NetworkTopology(links, flows, hosts, None)
+    network.write_to_json("test_case0_dummy.json")
```
name should be data/test_case_0_dummy.json.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+
+    # generate for FlowReno
+    flows = list()
+    f1 = FlowReno("F1", "H1", "H2", h1, h2, INITIAL_WINDOW_SIZE_PACKETS,
```
Again, named parameters and no references to objects like h1 and h2.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    f1 = FlowReno("F1", "H1", "H2", h1, h2, INITIAL_WINDOW_SIZE_PACKETS,
+                                  set(), list(), 20 * MEGABYTE, 1)
+    flows.append(f1)
+    h1.flows["F1"] = f1  # overwrites existing FlowDummy in h1
```
No references. These are set later in complete_initialization().

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    flows.append(f1)
+    h1.flows["F1"] = f1  # overwrites existing FlowDummy in h1
+    network = NetworkTopology(links, flows, hosts, None)
+    network.write_to_json("test_case0_reno.json")
```
data/test_case_0_reno.json

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+
+    # generate for FlowFast
+    flows = list()
+    f1 = FlowFast("F1", "H1", "H2", h1, h2, INITIAL_WINDOW_SIZE_PACKETS,
```
Same comments as before. These apply to the rest of the file too so I will not repeat them.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    flows.append(f1)
+    h1.flows["F1"] = f1
+    network = NetworkTopology(links, flows, hosts, None)
+    network.write_to_json("test_case0_fast.json")
```
data/test_case_0_fast.json

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    flows.append(f1)
+    h1.flows["F1"] = f1
+    network = NetworkTopology(links, flows, hosts, routers)
+    network.write_to_json("test_case1_dummy.json")
```
data/test_case_1_dummy.json

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    flows.append(f1)
+    h1.flows["F1"] = f1
+    network = NetworkTopology(links, flows, hosts, routers)
+    network.write_to_json("test_case1_reno.json")
```
data/test_case_1_reno.json

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    flows.append(f1)
+    h1.flows["F1"] = f1
+    network = NetworkTopology(links, flows, hosts, routers)
+    network.write_to_json("test_case1_fast.json")
```
data/test_case_1_fast.json

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    r4 = Router("R4")
+
+    links = list()
+    l0 = Link("H1", "R1", h1, r1, LinkBuffer(64 * KILOBYTE), 0.01, 12.5 * KILOBYTE, False)
```
Links cannot have their end Devices set. Also use named parameters everywhere as mentioned before.

lakshbhasin replied:
Rates are in _megabits_ per second, not kilobytes per second.

yubo56 replied:
Fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    # test case 0
+    links = list()
+    l1 = Link("H1", "H2", h1, h2, LinkBuffer(64 * KILOBYTE), 0.01,
+              10 * KILOBYTE, False)
```
Rates are in _megabits_ per second, not kilobytes per second.

lakshbhasin replied:
No need for setting default parameters like whether the link is busy.

yubo56 replied:
fixed

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,120 @@
+    # generate for FlowReno
+    flows = list()
+    f1 = FlowReno("F1", "H1", "H2", h1, h2, INITIAL_WINDOW_SIZE_PACKETS,
+                                  set(), list(), 20 * MEGABYTE, 1)
```
No need to set default parameters like set() and list() here.

yubo56 replied:
fixed

yubo56 committed:
Test case 0, 1 generation seem to be working

Test case 0, 1 generation works, need to test but have few random other
changes that cropped up, committng before further testing.
Updated .gitignore to ignore log, data directories

yubo56 committed:
Test Case 2, jsons are generating correctly

yubo56 committed:
Added more descriptive file header

yubo56 committed:
All jsons work in initializer now

Fixed various bugs in little places so now initializer is just waiting
for main event loops when running all the jsons. Jsons should be bug
free unless I linked the wrong Devices

lakshbhasin commented on network_simulator/common.py:
```diff
@@ -25,6 +25,11 @@
+"""Conversion to/from bits <-> KB, MB"""
+KILOBYTE = 8 * 1024
+MEGABYTE = 1024 * KILOBYTE
+MEGABIT = 128 * KILOBYTE
```
It's actually 125 (see Google). Megabit, kilobit, etc is base 10.

yubo56 replied:
fixed

lakshbhasin commented on network_simulator/common.py:
```diff
@@ -25,6 +25,11 @@
 ROUTER_PACKET_DEFAULT_ID = -1
 
+"""Conversion to/from bits <-> KB, MB"""
+KILOBYTE = 8 * 1024
```
Use floats whenever possible (8.0 vs 8) to make sure there's no nasty integer division bugs later.

yubo56 replied:
Fixed. If anybody ever needs integer division by KILOBYTE just cast to int.

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+    # test case 0
+    links = list()
+    l1 = Link(end_1_addr = "H1", end_2_addr = "H2", link_buffer =
+            LinkBuffer(max_buffer_size_bits = 64 * KILOBYTE),
```
Use floats for multipliers wherever possible (including below).

yubo56 replied:
done

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+    # generate for FlowDummy
+    flows = list()
+    f1 = FlowDummy(flow_id = "F1", source_addr = "H1", dest_addr = "H2",
+                                  data_size_bits = 20 * MEGABYTE,
```
Check that this indentation isn't underlined in PyCharm for failing PEP8 (grey underline). Generally a good idea to see what colors show up on the PyCharm sidebar before committing.

yubo56 replied:
PyCharm auto-fix imba

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+
+    # generate for FlowDummy
+    flows = list()
+    f1 = FlowDummy(flow_id = "F1", source_addr = "H1", dest_addr = "H2",
```
When using named parameters, don't have spaces around "=" signs. PyCharm should warn about this.

yubo56 replied:
done

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+if __name__ == "__main__":
+    # test case 0
+    links = list()
+    l1 = Link(end_1_addr = "H1", end_2_addr = "H2", link_buffer =
```
Can you add a "name" field to Link's constructor and a corresponding ivar? For example, the name of this link would be "L1". That would be useful for Sharon later on.

yubo56 replied:
Added. Default name is None (or should be ""?)

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+    links.append(l1)
+    links.append(l2)
+    links.append(l3)
+    lh1 = Link(end_1_addr = "S1", end_2_addr = "R1", link_buffer =
```
In this case, name the Links "LS1", "LS2", etc so we know what Hosts they connect to. Use PyCharm's refactoring option (SHIFT+F6 on non-Mac I think) to rename consistently.

yubo56 replied:
Um, which links do you think should be named LS#? The ones in test_case_2 only or test_case_{0,1} as well? I don't know what LS stands for so I don't know exactly what you mean haha

lakshbhasin commented:
After you make the changes I've mentioned and re-generate the JSONs (making sure they work with the initializer), this should be good to merge.

yubo56 committed:
Laksh's feedback, added name field to Link

KILOBYTE etc. in common.py now floats

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+
+    # router-router
+    links = list()
+    ls1 = Link(name="LS1", end_1_addr="R1", end_2_addr="R2",
```
name the router-router links L1, L2, L3 as in the test case.

lakshbhasin commented on test_case_gen.py:
```diff
@@ -0,0 +1,249 @@
+    links.append(ls1)
+    links.append(ls2)
+    links.append(ls3)
+    lh1 = Link(name="LH1", end_1_addr="S1", end_2_addr="R1",
```
Name the router-host link that connects to "S1" as "LS1". Similarly, the router-host link that connects to "T1" should be called "LT1", etc. Same for S2, S3, T2, and T3.

yubo56 replied:
renamed in hopefully sensible fashion

lakshbhasin commented on network_simulator/common.py:
```diff
@@ -25,6 +25,11 @@
 ROUTER_PACKET_DEFAULT_ID = -1
 
+"""Conversion to/from bits <-> KB, MB"""
+KILOBYTE = 8.0 * 1024
```
Actually, I believe kilobyte is actually 1000.0 bytes (i.e. 8000.0 bits) in networking. And a megabyte is 1000.0 kilobytes. See https://en.wikipedia.org/wiki/Megabyte for the distinction between megabyte and mebibyte. According to that Wikipedia link, "1 MB = 1000000 bytes ... is the definition recommended by the International System of Units (SI) and the International Electrotechnical Commission IEC.[2] This definition is used in networking contexts and most storage media."

You could also ask Cody to confirm (I'd recommend this).

yubo56 replied:
fixed, also asking the currently slumbering Cody

yubo56 committed:
Renamed some lynx

All links that are named in the specs are named as they are in the
specs. If unnamed in specs, are linked to a host, so named "L" +
hostname (e.g. if connected between H1, R1, named LH1).

Also fixed MB = 1000KB = 10^6B.

lakshbhasin commented:
Ok looks good to me (assuming the megabyte/megabit stuff is fine).
